### PR TITLE
[#1295] Add feature redirect if host does not exist

### PR DESCRIPTION
**Pull request by 030**
This allows to redirect to another URL instead of returning a 404

### Proposed changes
Describe the use case and detail of the change. If this PR addresses an issue on GitHub, make sure to include a link to that issue here in this description (not in the title of the PR).

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto master
- [x] I will ensure my PR is targeting the master branch and pulling from my branch from my own fork

### CODE REVIEW

**Comment by pleshakov** [comment]
Hi @030 

Thanks for the PR! We'll look into it and get back to you shortly.

**Comment by pleshakov** [comment]
Hi @030 

Most of the global configuration that affects the generated NGINX config we implement through the ConfigMap. This allows changing that configuration without re-creating the IC pods and also minimizes the number of cli args (and as a result helm chart parameters).

If this feature was implemented through the ConfigMap, I think the following ConfigMap key would be enough:

```yaml
default-server-redirect-url: https://nginx.org
```

At the same time, another way to solve this problem could be through a more general mechanism -- server snippets. We support snippets for Ingress resources via annotations and the ConfigMap. However, we don't support them for the default server.

For example, here we configure a redirect:
```yaml
default-server-server-snippets: location / { return 302 https://nginx.org; }
```

More use cases could be supported. Here we tell NGINX to close the connection:
```yaml
default-server-server-snippets: location / { return 444; }
```

Here we tell NGINX to return a static response: 
```yaml
default-server-server-snippets: location / { return 200 "welcome from the Ingress Controller!\n"; }
```

The default value would be  `location / { return 404; }`, for backward compatibilty.

I wonder what you think about the both ConfigMap approaches?

**Comment by 030** [comment]
@pleshakov Thank you for the feedback. Changes have been applied and the redirect works in conjunction with a configMap as well.

**Review by pleshakov — COMMENTED** [review]
Hi @030 

Thanks for addressing the feedback.

Because in the implementation `default-server-redirect-url` configures the `return` directive, could you rename the key to `default-server-return` ? Note that the supported use cases of the return directive are broader than redirecting. That's why `default-server-return` suites better.  Could you also make the corresponding changes to the identifiers in the code?

In my original suggestion, I assumed that `default-server-redirect-url` would only configures the redirect. For example, `return 302 {{.DefaultServerRedirectURL}};`.  However, current the implementation with `return {{.DefaultServerRedirectURL};` makes sense and covers more use cases. So I think it is better to keep it, although with the mentioned renaming.

There are also a few additional comments below.

**Comment by 030** [comment]
@pleshakov Variable has been renamed and the documentation has been moved to the general customization paragraph

File: docs-web/configuration/global-configuration/configmap-resource.md [hunk 1]
```diff
@@ -114,6 +114,10 @@ See the doc about [VirtualServer and VirtualServerRoute resources](/nginx-ingres
      - Enables or disables the `real_ip_recursive <https://nginx.org/en/docs/http/ngx_http_realip_module.html#real_ip_recursive>`_ directive.
      - ``False``
      - 
+   * - ``default-server-return``
+     - Configures the `return <https://nginx.org/en/docs/http/ngx_http_rewrite_module.html#return>`_  directive in the default server, which handles a client request if none of the hosts of Ingress or VirtualServer resources match. The default value configures NGINX to return a 404 error page. You can configure a fixed response or a redirect. For example,  ``default-server-return: 302 https://nginx.org`` will redirect a client to https://nginx.org.
```
Comment (jputrino) [inline docs-web/configuration/global-configuration/configmap-resource.md:118]:
```suggestion
     - Configures the `return <https://nginx.org/en/docs/http/ngx_http_rewrite_module.html#return>`_  directive in the default server, which handles a client request if none of the hosts of Ingress or VirtualServer resources match. The default value configures NGINX to return a 404 error page. You can configure a fixed response or a redirect. For example, ``default-server-return: 302 https://nginx.org`` will redirect a client to https://nginx.org.
```

**Review by pleshakov — APPROVED** [review]
Hi @030 

This looks good to me 👍  (with this comment applied https://github.com/nginxinc/kubernetes-ingress/pull/1297/files#r560548874 )

If you could also update the commit message, as it still refers to the redirect feature. Also, for consistency, could you remove the `[#1295]` from the beginning of a commit message? we typically don't put the issue number in the title -- https://github.com/nginxinc/kubernetes-ingress/commits/master

I will ask for one more approval from our project maintainers.

**Review by Dean-Coakley — APPROVED** [review]
Thanks very much for your continued efforts on this @030 !

LGTM.

This looks ready to be merged once the feedback from @jputrino is resolved.

**Pull request merged**